# GDELT Exploration

## Objective

To explore the GDELT Events dataset, understand its structure,
identify available features, and evaluate its potential usefulness
for our research problem.

In [1]:
import pandas as pd
import requests
import matplotlib.pyplot as plt

In [2]:
import requests
import time
import pandas as pd

url = "https://api.gdeltproject.org/api/v2/doc/doc"

params = {
    "query": "India",
    "mode": "ArtList",
    "maxrecords": 10,
    "format": "json"
}

time.sleep(10)

response = requests.get(url, params=params)

print("Status code:", response.status_code)

Status code: 429


In [3]:
print(response.text[:1000])

Please limit requests to one every 5 seconds or contact kalev.leetaru5@gmail.com for larger queries. All high-traffic users should switch to our ngrams dataset: https://blog.gdeltproject.org/using-the-new-web-ngrams-dataset-to-find-relevant-coverage/. For trend analysis, please see our daily newsletter briefings: https://blog.gdeltproject.org/announcing-our-new-daily-todays-trends-on-capitol-hill-todays-media-trends-newsletter-briefings/.




## Experiment 1: Disaster-Related News Retrieval

The objective of this experiment is to evaluate how GDELT can be used as a source of news data for disaster-related research.

Four disaster categories are considered:

- Flood
- Earthquake
- Cyclone
- Wildfire

For each category, news articles related to India will be retrieved and compared based on the number of articles and available metadata.

In [4]:
import requests
import pandas as pd

### Defining the Disaster Queries

The following queries are used to retrieve news articles related to major types of disasters in India.

In [5]:
disaster_queries = {
    "Flood": "flood India",
    "Earthquake": "earthquake India",
    "Cyclone": "cyclone India",
    "Wildfire": "wildfire India"
}

disaster_queries

{'Flood': 'flood India',
 'Earthquake': 'earthquake India',
 'Cyclone': 'cyclone India',
 'Wildfire': 'wildfire India'}

### Retrieving Flood-Related News

The first experiment focuses on flood-related news in India.

The GDELT DOC 2.0 API is queried for recent articles matching the search term `flood India`.

In [9]:
import time

time.sleep(10)

url = "https://api.gdeltproject.org/api/v2/doc/doc"

params = {
    "query": disaster_queries["Flood"],
    "mode": "artlist",
    "format": "json",
    "maxrecords": 20,
    "timespan": "7d"
}

response = requests.get(url, params=params, timeout=30)

print("Status Code:", response.status_code)

Status Code: 429


In [10]:
if response.status_code == 200:
    data = response.json()
    print("Number of articles returned:", len(data.get("articles", [])))
else:
    print("Request failed.")
    print("Status Code:", response.status_code)
    print("Response:", response.text[:500])

Request failed.
Status Code: 429
Response: Please limit requests to one every 5 seconds or contact kalev.leetaru5@gmail.com for larger queries. All high-traffic users should switch to our ngrams dataset: https://blog.gdeltproject.org/using-the-new-web-ngrams-dataset-to-find-relevant-coverage/. For trend analysis, please see our daily newsletter briefings: https://blog.gdeltproject.org/announcing-our-new-daily-todays-trends-on-capitol-hill-todays-media-trends-newsletter-briefings/.




In [11]:
print(response.headers)

{'Date': 'Sat, 19 Sep 2026 06:46:47 GMT', 'Server': 'GDELT Server', 'Keep-Alive': 'timeout=5, max=20', 'Connection': 'Keep-Alive', 'Transfer-Encoding': 'chunked'}


In [12]:
print(response.text[:1000])

Please limit requests to one every 5 seconds or contact kalev.leetaru5@gmail.com for larger queries. All high-traffic users should switch to our ngrams dataset: https://blog.gdeltproject.org/using-the-new-web-ngrams-dataset-to-find-relevant-coverage/. For trend analysis, please see our daily newsletter briefings: https://blog.gdeltproject.org/announcing-our-new-daily-todays-trends-on-capitol-hill-todays-media-trends-newsletter-briefings/.




## Observation: API Rate Limiting

During the experiment, the GDELT DOC 2.0 API returned HTTP status code 429
(Too Many Requests). The server response indicated that requests should be
limited to one request every 5 seconds.

This shows that GDELT applies request-rate limitations to API users.
Therefore, API requests need to be controlled carefully when conducting
programmatic data collection.

In [13]:
print([name for name in globals() if not name.startswith("_")])

['In', 'Out', 'get_ipython', 'exit', 'quit', 'open', 'pd', 'requests', 'plt', 'time', 'url', 'params', 'response', 'disaster_queries']


## Experiment 2: Retrieving and Structuring Disaster News Data

After identifying the API rate limitation, the next step is to retrieve a
small set of disaster-related news articles and convert the returned
data into a structured Pandas DataFrame for analysis.

In [15]:
import time
import requests
import pandas as pd

# Wait before making a new request
time.sleep(10)

url = "https://api.gdeltproject.org/api/v2/doc/doc"

params = {
    "query": "flood India",
    "mode": "artlist",
    "format": "json",
    "maxrecords": 10,
    "timespan": "7d"
}

response = requests.get(url, params=params, timeout=30)

print("Status Code:", response.status_code)

Status Code: 429


### Result of Experiment 2

The attempt to retrieve a new disaster-related dataset using the GDELT
DOC 2.0 API returned HTTP status code 429 (Too Many Requests).

The server message indicated that API requests should be limited to one
request every five seconds. Even after introducing a delay in the Python
request, the API continued to return the rate-limit response.

Therefore, a new disaster dataset could not be retrieved during this
testing session. The result was recorded as a practical limitation of
using the GDELT API.

## Preliminary Findings

GDELT was explored as a potential source of news and media data for
disaster-related research. The GDELT DOC 2.0 API provides search-based
access to news coverage and supports structured outputs such as JSON,
which can be used for programmatic analysis.

During the practical experiment, API requests were successfully sent to
the GDELT server, but subsequent requests returned HTTP 429 (Too Many
Requests). The server indicated that requests should be limited to one
request every five seconds.

Therefore, the experiment demonstrated both the potential usefulness of
GDELT as a news-data source and an important practical limitation:
programmatic data collection must account for API rate limiting.

Further analysis will be performed only after obtaining a successful
dataset or through an alternative GDELT data-access method.